### PR TITLE
Refactor QAT to use tensor subclasses

### PR DESCRIPTION
**Pull request by andrewor14**
## Overview

This commit refactors QAT to use tensor subclasses. This is motivated by the general move towards tensor subclasses in torchao for better composability with other subclasses like DTensors. To achieve this, we introduce `AffineFakeQuantizedTensor`, which is analogous to `AffineQuantizedTensor` but applies fake quantization instead and requires gradient updates.

## How training works
`AffineFakeQuantizedTensor` wraps the original weight or input activation tensor and applies fake quantize dynamically only when the linear function is called. Gradients only flow to the outer tensor (`AffineFakeQuantizedTensor`) and never to the inner tensor. For weights, the outer tensor is also a `torch.nn.Parameter`, and gradient updates received by the outer tensor are then passed to the inner tensor through ops like `aten.add_` and `aten.mul_`.

## Input activation

An important difference between the PTQ and the QAT flows is how input activation subclasses are inserted. For QAT, we use the nn.module `forward_pre_hook` instead of relying on another subclass `LinearActivationQuantizedTensor` that wraps the weight subclass. The problem with the old PTQ approach is it can create subclasses under `__torch_dispatch__`, which runs below autograd and so the created subclasses cannot have gradients, so it was difficult to get the gradients to flow correctly in such cases. It's also not super intuitive because quantizing input activation needs to go through the weights. In the new approach used by QAT, we instead register a `forward_pre_hook` that wraps the input activations before each call to forward. This approach is also motivated by how [DTensor wraps their subclasses ](https://github.com/pytorch/pytorch/blob/844103197d3e8cf6b4b59176e473365113f4f962/torch/distributed/tensor/parallel/style.py#L521). 

## Work items

- [x] Add AffineFakeQuantizedTensor
- [x] Add support for int4 weight only fake quantize
- [x] Add support for int8 dynamic activations + int4 weight fake quantize (8da4w)
- [x] Add prepare and convert path to int4 QAT quantizer
- [x] Add prepare and convert path to 8da4w QAT quantizer
- [x] Support enabling and disabling fake quant dynamically
- [x] Support `__repr__` in AffineFakeQuantizedTensor
- [x] Fix backward pass for int4 weight only
- [x] Fix backward pass for int8 dynamic activations + int4 weight

### CODE REVIEW

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        return torch.Tensor._make_wrapper_subclass(cls, original_tensor.shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        original_tensor: torch.Tensor,
+        apply_fake_quant_fn: Callable,
+        fake_quant_enabled: bool = True,
+    ):
+        # TODO: original_tensor is not getting updated!
+        original_tensor.requires_grad_(self.requires_grad)
```
Comment (andrewor14) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:57]: here

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    _implements,
+    _dispatch__torch_function__,
+    _dispatch__torch_dispatch__,
+)
+from .utils import _GenericFakeQuantize
+
+aten = torch.ops.aten
+
+
+class _FromTorchTensor(torch.autograd.Function):
```
Comment (andrewor14) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:21]: here

File: torchao/quantization/prototype/qat/api.py [hunk 3]
```diff
@@ -18,20 +18,88 @@
...
+
+        def input_quant_func(x: torch.Tensor):
+            return to_affine_fake_quantized(
+                x,
+                input_mapping_type,
+                _get_per_token_block_size(x),
+                input_target_dtype,
+            )
+
+        weight = to_affine_fake_quantized(
```
Comment (bdhirsh) [inline torchao/quantization/prototype/qat/api.py:89]: if your only goal is "my model has an `m.weight` is a tensor subclass, I want to be able to run the model, call `.backward()`, and get gradient computed directly on my subclass", then I think instead of this all you really need is:

```
weight = MySubclass(inner_tensor)
weight.requires_grad_(True)
model.weight = torch.nn.Parameter(weight)
```

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        return torch.Tensor._make_wrapper_subclass(cls, original_tensor.shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        original_tensor: torch.Tensor,
+        apply_fake_quant_fn: Callable,
+        fake_quant_enabled: bool = True,
+    ):
+        # TODO: original_tensor is not getting updated!
+        original_tensor.requires_grad_(True)
```
Comment (andrewor14) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:93]: here

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    def __new__(
+        cls,
+        original_tensor: torch.Tensor,
+        apply_fake_quant_fn: Callable,
+        fake_quant_enabled: bool = True,
+    ):
+        kwargs = {}
+        kwargs["device"] = original_tensor.device
+        kwargs["dtype"] = original_tensor.dtype
+        kwargs["requires_grad"] = True
```
Comment (andrewor14) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:83]: here

File: torchao/quantization/quant_api.py [hunk 6]
```diff
@@ -257,9 +259,9 @@ def replace_conv2d_1x1(conv):
     )
 
 
-def _get_linear_subclass_inserter(constructor):
+def _get_linear_subclass_inserter(constructor, requires_grad=False):
     def insert_subclass(lin):
-        lin.weight = torch.nn.Parameter(constructor(lin.weight), requires_grad=False)
+        lin.weight = torch.nn.Parameter(constructor(lin.weight), requires_grad=requires_grad)
```
Comment (andrewor14) [inline torchao/quantization/quant_api.py:264]: here

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+                qmin,
+                qmax,
+                eps,
+                scale_dtype,
+                zero_point_dtype,
+                preserve_zero,
+                zero_point_domain,
+            )
+            fq = _GenericFakeQuantize.apply(
+                t,
```
Comment (andrewor14) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:124]: here

File: torchao/quantization/prototype/qat/utils.py [hunk 8]
```diff
@@ -143,3 +179,60 @@ def _choose_qparams_per_token_asymmetric(
...
+
+def _forward_pre_hook_handler(
+    mod: torch.nn.Linear,
+    prehook: Callable,
+    handler: torch.utils.hooks.RemovableHandle,
+):
+    """
+    Store a 2-tuple (prehook function, handler) as an attribute on the given linear module.
+    """
+    setattr(mod, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK, (prehook, handler))
```
Comment (bdhirsh) [inline torchao/quantization/prototype/qat/utils.py:191]: why do you need to stash the hooks / handles themselves onto the module? (when you register an nn module hook, it will automatically be stored somewhere on the module)

Comment (bdhirsh) [reply torchao/quantization/prototype/qat/utils.py:191]: oh I see, you want the user to be able to enable/disable the entire flow on a given layer.

File: torchao/quantization/prototype/qat/utils.py [hunk 9]
```diff
@@ -29,19 +38,21 @@ class _GenericFakeQuantize(torch.autograd.Function):
...
-        assert scales.dtype == torch.float32
-        assert zero_points.dtype == torch.int32
+        # avoid circular dependencies
+        from torchao.quantization.prototype.qat.affine_fake_quantized_tensor import (
+            AffineFakeQuantizedTensor,
+        )
+
+        original_input = input
+        if isinstance(original_input, AffineFakeQuantizedTensor):
+            input = input.original_tensor
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/utils.py:55]: why don't we pass in original_tensor directly here?

File: test/quantization/test_qat.py [hunk 10]
```diff
@@ -373,87 +399,9 @@ def _assert_close_4w(self, val, ref):
         print(mean_err)
         self.assertTrue(mean_err < 0.05)
 
+    # TODO: compare against quantize_ API instead
```
Comment (jerryzh168) [inline test/quantization/test_qat.py:402]: nit: is this already done for this test case

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+                qmin,
+                qmax,
+                zero_point_domain,
+            )
+            return fq
+        return cls(
+            input_float,
+            apply_fake_quant_fn,
+            fake_quant_enabled=True,
+            requires_grad=True,
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:128]: should we remove this and just get this from `input_float.requires_grad`?

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        args[2] if len(args) > 2 else None,
+    )
+    if isinstance(input_tensor, AffineFakeQuantizedTensor):
+        input_tensor = input_tensor.get_value()
+    if isinstance(weight_tensor, AffineFakeQuantizedTensor):
+        weight_tensor = weight_tensor.get_value()
+    return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+
+@implements([aten.mm.default, aten.addmm.default])
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:197]: nit: might be cleaner just to split the implementation for these two ops

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+@implements([
+    aten.add.Tensor,
+    aten.add_.Tensor,
+    aten.mul_.Tensor,
+])
+def _(func, types, args, kwargs):
+    assert len(args) == 2, f"dispatched the wrong op to the binary handler: {func}"
+    new_args = pytree.tree_map_only(AffineFakeQuantizedTensor, lambda x: x.original_tensor, args)
+    first_afq_tensor = args[0] if isinstance(args[0], AffineFakeQuantizedTensor) else args[1]
+    fn = lambda x: func(*new_args, **kwargs)
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:247]: it seems that `x` is ignored, looks weird

File: torchao/quantization/quant_api.py [hunk 14]
```diff
@@ -258,9 +260,30 @@ def replace_conv2d_1x1(conv):
...
+        requires_grad: whether the weight parameter requires gradients (defaults to False)
+    """
     def insert_subclass(lin):
-        lin.weight = torch.nn.Parameter(constructor(lin.weight), requires_grad=False)
+        lin.weight = torch.nn.Parameter(weight_constructor(lin.weight), requires_grad=requires_grad)
+        if input_constructor is not None:
+            prehook = lambda _, args: tuple([input_constructor(args[0])] + list(args[1:]))
+            handle = lin.register_forward_pre_hook(prehook)
+            if forward_pre_hook_handler is not None:
+                forward_pre_hook_handler(lin, prehook, handle)
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:286]: this looks a bit complicated, do we need to create a general helper function? what about just have some function specific for QAT for now

File: torchao/quantization/prototype/qat/api.py [hunk 15]
```diff
@@ -194,23 +232,54 @@ def _get_qmin_qmax(self, n_bit: int):
...
+            block_size,
+            target_dtype,
+            quant_min,
+            quant_max,
+            eps,
+            zero_point_dtype=zero_point_dtype,
+            preserve_zero=preserve_zero,
+            zero_point_domain=zero_point_domain,
+        )
+    return _get_linear_subclass_inserter(_apply_fake_quant, requires_grad=True)
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/api.py:281]: I feel we could also get `requires_grad` from weight directly?

File: torchao/quantization/prototype/qat/utils.py [hunk 16]
```diff
@@ -143,3 +179,61 @@ def _choose_qparams_per_token_asymmetric(
...
+    assert isinstance(weight, AffineFakeQuantizedTensor)
+    weight.fake_quant_enabled = enable
+
+    # Enable/disable input fake quant
+    if hasattr(mod, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK):
+        (prehook, handle) = getattr(mod, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK)
+        if enable:
+            if handle is None:
+                handle = mod.register_forward_pre_hook(prehook)
+                _forward_pre_hook_handler(mod, prehook, handle)
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/utils.py:236]: I assume this is because we are passing `_forward_pre_hook_handler` to `_get_linear_subclass_inserter`, maybe we can clean this up if we just create a helper function just for QAT

File: torchao/quantization/prototype/qat/utils.py (hunk 9 above)
Comment (andrewor14) [reply torchao/quantization/prototype/qat/utils.py:55]: We're passing in the `original_tensor` here. Maybe the naming is confusing? Let me rewrite this

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py (hunk 11 above)
Comment (andrewor14) [reply torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:128]: I feel it's better to always set it to True here since we always want gradients for QAT

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py (hunk 13 above)
Comment (andrewor14) [reply torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py]: I think the problem is for these ops, I found that sometimes args[1] is the subclass, so we need to handle both cases. There's a bit of complexity here so I feel it's better to do it in a common function

Comment (andrewor14) [reply torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py]: I rewrote this a bit to use a new helper function that simply accepts a new value, hope that's clearer

File: torchao/quantization/prototype/qat/utils.py (hunk 16 above)
Comment (andrewor14) [reply torchao/quantization/prototype/qat/utils.py:237]: done

File: test/quantization/test_qat.py [hunk 17]
```diff
@@ -247,6 +266,16 @@ def test_qat_8da4w_quantizer_disable_fake_quant(self):
             enable_8da4w_fake_quant,
         )
 
+        def assert_fake_quant_enabled(m: torch.nn.Linear, enabled: bool):
+            self.assertTrue(isinstance(m.weight, AffineFakeQuantizedTensor))
+            self.assertEqual(m.weight.fake_quant_enabled, enabled)
+            self.assertTrue(hasattr(m, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK))
+            (_, handle) = getattr(m, _QAT_LINEAR_SUBCLASS_INPUT_PREHOOK)
```
Comment (jerryzh168) [inline test/quantization/test_qat.py:273]: btw, this feels like an implementation detail, so ideally we can check through other ways like check if the gradient of weight is updated

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py (hunk 11 above)
Comment (jerryzh168) [reply torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:128]: even that I feel just reusing the requires_grad from `input_float` might be beneficial, otherwise we'd have to worry about the inconsistency here, similar for other things like dtype, shape, device etc.

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+    def _create_new(self, new_value: torch.Tensor):
+        """
+        Create a new `AffineFakeQuantizedTensor` with a new value,
+        to be called within __torch_dispatch__.
+        """
+        return self.__class__(
+            new_value,
+            self.apply_fake_quant_fn,
+            self.fake_quant_enabled,
+            requires_grad=False,
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:180]: so why is this set to `False` here? maybe adding a bit more detail in the docstring might help

File: torchao/quantization/prototype/qat/utils.py [hunk 19]
```diff
@@ -143,3 +180,83 @@ def _choose_qparams_per_token_asymmetric(
     zero_point = torch.clamp(zero_point, qmin, qmax).round()
 
     return scale.to(scales_precision), zero_point.to(zero_points_precision)
+
+def _forward_pre_hook_handler(
```
Comment (jerryzh168) [inline torchao/quantization/prototype/qat/utils.py:184]: maybe we can remove this helper function now and just inline the changes

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py (hunk 18 above)
Comment (andrewor14) [reply torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py:180]: Yeah sure, will explain in more detail in the docstring. It's because this is called in `__torch_dispatch__`, which cannot produce tensors with gradients (because autograd will try to attach metadata after `__torch_dispatch__` and will error if it finds existing metadata).

File: test/quantization/test_qat.py (hunk 17 above)
Comment (andrewor14) [reply test/quantization/test_qat.py:273]: yeah I'll add a separate test for that

Comment (andrewor14) [reply test/quantization/test_qat.py:273]: yeah sure, I'll add a separate test for that. I feel it's still good to test that the prehook was installed correctly though

File: torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py (hunk 11 above)
Comment (andrewor14) [reply torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py]: Ok, relaxed this to use the original tensor's `requires_grad`

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks!

**Comment by andrewor14** [comment]
For reference, just added some end-to-end benchmarks and evaluation results for tensor subclass QAT here: https://github.com/pytorch/torchtune/pull/1330. Looks pretty good so far
